mlflow/xgboost.py

```diff
@@ -416,10 +437,36 @@ def log_feature_importance_plot(features, importance, importance_type):
             finally:
                 shutil.rmtree(tmpdir)
 
-        try_mlflow_log(log_model, model, artifact_path="model")
+        # dtrain must exist as the original train function already ran successfully
+        dtrain = args[1] if len(args) > 1 else kwargs["dtrain"]
+
+        input_example = None
+        signature = None
+        try:
+            input_example = dtrain.data_copy
+
+            # it is possible that the dataset was constructed before the patched
+            #   constructor was applied, so we cannot assume the data_copy exists
+            if input_example is None:
+                raise Exception("failed to gather example input.")
+
+            model_output = model.predict(xgb.DMatrix(input_example))
+            signature = infer_signature(input_example, model_output)
+        except Exception as e:  # pylint: disable=broad-except
+            msg = "Failed to infer the model signature: " + str(e)
```

Comment:
In the case where `input_example` is assigned to a non-None value but `infer_signature` fails, we should reset `input_example` to `None` before passing it to `log_model()`. I noticed we have a test case for this (`test_xgb_autolog_continues_logging_even_if_signature_inference_fails`), but it doesn't seem to assert that `input_example` is not present when signature inference fails. We should add that for consistency with the behavior of `mlflow.skearn.autolog()` and `mlflow.lightgbm.autolog()`.

Though it does occur to me that `input_example` may be independently useful even if `signature` cannot be inferred. @harupy Can you weigh in here? I forgot why we decided to reset `input_example` in this case.

Reply:
@dbczumar 

> I forgot why we decided to reset input_example in this case.

I remember it's for simplicity.

Reply:
Found a related comment: https://github.com/mlflow/mlflow/pull/3326#discussion_r478742287